server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java

```diff
@@ -0,0 +1,584 @@
+import org.apache.accumulo.server.conf.store.PropChangeListener;
+import org.apache.accumulo.server.conf.store.PropStore;
+import org.apache.accumulo.server.conf.store.PropStoreException;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.benmanes.caffeine.cache.Ticker;
+import com.google.common.annotations.VisibleForTesting;
+
+public class ZooPropStore implements PropStore, PropChangeListener {
+
+  private final static Logger log = LoggerFactory.getLogger(ZooPropStore.class);
+  private final static VersionedPropCodec codec = VersionedPropGzipCodec.codec(true);
+
+  private final ZooReaderWriter zrw;
+  private final PropStoreWatcher propStoreWatcher;
+  private final PropCache cache;
+  private final PropStoreMetrics cacheMetrics = new PropStoreMetrics();
+  private final ReadyMonitor zkReadyMon;
+
+  /**
+   * Create instance using ZooPropStore.Builder
+   *
+   * @param instanceId
+   *          the instance id
+   * @param zrw
+   *          a wrapper set of utilities for accessing ZooKeeper.
+   * @param readyMonitor
+   *          coordination utility for ZooKeeper connection status.
+   * @param propStoreWatcher
+   *          an extended ZooKeeper watcher
+   * @param ticker
+   *          a synthetic clock used for testing.
+   */
+  private ZooPropStore(final InstanceId instanceId, final ZooReaderWriter zrw,
+      final ReadyMonitor readyMonitor, final PropStoreWatcher propStoreWatcher,
+      final Ticker ticker) {
+
+    this.zrw = zrw;
+    this.zkReadyMon = readyMonitor;
+    this.propStoreWatcher = propStoreWatcher;
+
+    MetricsUtil.initializeProducers(cacheMetrics);
+
+    ZooPropLoader propLoader = new ZooPropLoader(zrw, codec, propStoreWatcher, cacheMetrics);
+
+    if (ticker == null) {
+      cache = new CaffeineCache.Builder(propLoader, cacheMetrics).build();
+    } else {
+      cache = new CaffeineCache.Builder(propLoader, cacheMetrics).withTicker(ticker).build();
+    }
+
+    try {
+      var path = ZooUtil.getRoot(instanceId);
+      if (zrw.exists(path, propStoreWatcher)) {
+        log.debug("Have a ZooKeeper connection and found instance node: {}", instanceId);
+        zkReadyMon.setReady();
+      } else {
+        throw new IllegalStateException("Instance may not have been initialized, root node: " + path
+            + " does not exist in ZooKeeper");
+      }
+    } catch (InterruptedException | KeeperException ex) {
+      throw new IllegalStateException("Failed to read root node " + instanceId + " from ZooKeeper",
+          ex);
+    }
+  }
+
+  public static VersionedPropCodec getCodec() {
+    return codec;
+  }
+
+  public static PropStore initialize(final InstanceId instanceId, final ZooReaderWriter zrw) {
+    return new ZooPropStore.Builder(instanceId, zrw, zrw.getSessionTimeout()).build();
```

Comment:
Multiple instances aren't expected to be started, but we don't want to design things statically. We're trying to move away from that, because it prevents VM reuse, causes issues with debuggers and application servers, or other containerized methods for running processes. The more we give objects a concrete lifecycle, rather than create static singletons, the better we can manage our resources.

Reply:
Ok, I understand that, but we are not doing anything to protect ourselves here. There is nothing to prevent multiple instances of these objects from being created in the future except a code review. Could we at least make the Builder methods `private`, remove the `public` modifier from this method, and move this class to the same package from where it's called?

Reply:
https://github.com/apache/accumulo/issues/2635 to look at this as a follow-on